src/session-adapters/internalAdapter.ts

```diff
@@ -1,54 +1,148 @@
 import http from 'http'
 import { Adapter, MaybePromise } from './generalAdapter'
+import {
+  CallbackCookieData,
+  callbackCookieName,
+} from '../storyblok-auth-api/handle-requests/callbackCookie'
+import {
+  expireCookie,
+  getCookie,
+  setCookie,
+  signData,
+  verifyData,
+} from '../utils'
+import { AppSession } from '../session/types'
+import { AuthHandlerParams } from '../storyblok-auth-api'
 
 export type InternalAdapter = {
-  getItem: (key: string) => MaybePromise<string | object | undefined>
-  setItem: (params: {
-    key: string
-    value: string | object
-  }) => MaybePromise<void>
-  removeItem: (key: string) => MaybePromise<void>
-  hasItem: (key: string) => MaybePromise<boolean>
+  // session
+  getSession: (params: {
+    spaceId: string
+    userId: string
+  }) => MaybePromise<AppSession | undefined>
+
+  setSession: (params: {
+    spaceId: string
+    userId: string
+    session: AppSession
+  }) => MaybePromise<boolean>
+
+  removeSession: (params: {
+    spaceId: string
+    userId: string
+  }) => MaybePromise<boolean>
+
+  hasSession: (params: {
+    spaceId: string
+    userId: string
+  }) => MaybePromise<boolean>
+
+  // callback data
+  setCallbackData: (data: CallbackCookieData) => MaybePromise<boolean>
+
+  getCallbackData: () => MaybePromise<CallbackCookieData | undefined>
+
+  removeCallbackData: () => MaybePromise<boolean>
 }
 
 type CreateInternalAdapter = ({
+  params,
   req,
   res,
   adapter,
 }: {
+  params: Pick<AuthHandlerParams, 'clientId' | 'clientSecret' | 'cookieName'>
   req: http.IncomingMessage
   res: http.ServerResponse
   adapter: Adapter
 }) => InternalAdapter
 
 export const createInternalAdapter: CreateInternalAdapter = ({
+  params,
   req,
   res,
   adapter,
-}) => ({
-  getItem: (key) =>
-    adapter.getItem({
-      req,
-      res,
-      key,
-    }),
-  setItem: ({ key, value }) =>
-    adapter.setItem({
-      req,
-      res,
-      key,
-      value,
-    }),
-  hasItem: (key) =>
-    adapter.hasItem({
-      req,
-      res,
-      key,
-    }),
-  removeItem: (key) =>
-    adapter.removeItem({
-      req,
-      res,
-      key,
-    }),
-})
+}) => {
+  const sessionKey = params.cookieName || 'sb.auth'
+
+  return {
+    getSession: async ({ spaceId, userId }) => {
+      const session = await adapter.getItem({
+        req,
+        res,
+        clientId: params.clientId,
+        spaceId,
+        userId,
+        key: sessionKey,
+      })
+      if (!session) {
+        return undefined
+      }
+      try {
+        return JSON.parse(session) as AppSession
+      } catch (err) {
+        return undefined
+      }
+    },
+
+    setSession: async ({ spaceId, userId, session }) => {
```

Comment:
In this signature do you think we need spaceId and userId? because the session should have both right?

Reply:
yeah I had the same thought. but I feel like having spaceId and userId explicitly makes it clear that it sets the session based on spaceId and userId. What do you think?

Reply:
I see your point, I do not have a strong preference here. So let's leave as is :)